aten/src/THC/THCIntegerDivider.cuh

```diff
@@ -0,0 +1,120 @@
+// <= m < 2^(N+1)) and shift s such that:
+//
+//    \floor(n / d) = \floor((m * n) / 2^(N+s)).
+//
+// Given such m and s, the integer division can be then implemented as:
+//
+//    let m' = m - 2^N  // 0 <= m' < 2^N
+//        s1 = (s == 0) ? 0 : 1
+//        s2 = s - s1
+//
+//    fast_integer_division(n):
+//      // Multiply two N-bit unsigned integers: the result is a 2N-bit unsigned
+//      // integer.  Then take the higher N bits.
+//      t = (m' * n) >> N
+//
+//      // The return value is simply (t + n) >> s, but computing this directly
+//      // may overflow, so we shift twice.
+//      t2 = t + ((n - t) >> s1)
+//      return t2 >> s2
+//
+// We have to be careful with (t + n) because it may overflow.  Finding such a
+// magic number is surprisingly easy:
+//
+//    s  = \ceil(\log_2 d)
+//    m' = \floor(2^N * (2^s - d) / d) + 1  // Need 2N-bit integer arithmetic.
+//
+// See also:
+//    - Division by Invariant Integers Using Multiplication,
+//      Torbjörn Granlund and Peter L. Montgomery, 1994.
+//
+//    - http://www.hackersdelight.org/magic.htm
+//
+//    - http://ridiculousfish.com/blog/posts/labor-of-division-episode-i.html
+
+// Result of div/mod operation stored together.
+template <typename Value>
+struct DivMod {
+  Value div, mod;
+
+  __host__ __device__ DivMod(Value div, Value mod) : div(div), mod(mod) { }
+};
+
+// Base case: we only have an implementation for uint32_t for now.  For
+// everything else, we use plain division.
+template <typename Value>
+struct IntDivider {
+  IntDivider() { }  // Dummy constructor for arrays.
+  IntDivider(Value d) : divisor(d) { }
+
+  __host__ __device__ inline Value div(Value n) const { return n / divisor; }
+  __host__ __device__ inline Value mod(Value n) const { return n % divisor; }
+  __host__ __device__ inline DivMod<Value> divmod(Value n) const {
+    return DivMod<Value>(n / divisor, n % divisor);
+  }
+
+  Value divisor;
+};
+
+// Implement fast integer division.
+template <>
+struct IntDivider<unsigned int> {
+  static_assert(sizeof(unsigned int) == 4, "Assumes 32-bit unsigned int.");
+
+  IntDivider() { }  // Dummy constructor for arrays.
+
+  IntDivider(unsigned int d) : divisor(d) {
+    assert(divisor != 0);
+
+    // TODO: gcc/clang has __builtin_clz() but it's not portable.
+    unsigned int shift;
+    for (shift = 0; shift < 32; shift++) if ((1U << shift) >= divisor) break;
+    s1 = (shift == 0) ? 0 : 1;
+    s2 = shift - s1;
+
+    uint64_t one = 1;
+    uint64_t magic = ((one << 32) * ((one << shift) - divisor)) / divisor + 1;
+    m1 = magic;
+    assert(m1 > 0 && m1 == magic);  // m1 must fit in 32 bits.
+  }
+
+  __host__ __device__ inline unsigned int div(unsigned int n) const {
+#ifdef __CUDA_ARCH__
+    // 't' is the higher 32-bits of unsigned 32-bit multiplication of 'n' and
+    // 'm1'.
+    unsigned int t = __umulhi(n, m1);
+    unsigned int t2 = t + ((n - t) >> s1);
```

Comment:
I changed `canUse32BitIndexMath` so that dividend/divisor are at most INT32_MAX.  Given the reduced range, the code can now work with `__umulhi`, one addition, and one shift.

(The Caffe2 code gets rid of the addition, but I think that requires 64-bit math.  An addition is probably better than adding a conditional check for `divisor==1`.)